greedy_max/schedule.py

```diff
@@ -81,30 +143,26 @@ def __init__(self,
         self.fpu_to_barcode = fpu2b
         self.ifu = ifus
 
-    # TODO: This method can be static and uses no data from TimeSlots.
-    # TODO: No idea what these variables are. ni vs nint?
-    def intervals(self, empty_slots: np.ndarray) -> np.ndarray:
-        ni = len(empty_slots)
-        cvec = np.zeros(ni, dtype=int)
-        nint = 1
-        cvec[0] = nint
-        for j in range(1, ni):
-            if empty_slots[j] != (empty_slots[j - 1] + 1):
-                nint = nint + 1
-            cvec[j] = nint 
-
-        idx = np.digitize(cvec, bins=np.arange(ni) + 1)
-
-        return idx 
+    def non_zero_intervals(self, site: Site, obs_idx: int, interval: np.ndarray) -> np.ndarray:
 
+        weights_on_interval = self.weights[site][obs_idx][interval]
+        # Create an array that is 1 where the weights is greater than 0, and pad each end with an extra 0.
+        isntzero = np.concatenate(([0], np.greater(weights_on_interval, 0), [0]))
+        absdiff = np.abs(np.diff(isntzero))
+        # Get the ranges for each non zero interval
+        ranges = np.where(absdiff == 1)[0].reshape(-1, 2) 
```

Comment:
Little confused by what this does but that may be my numpy lack of knowledge showing through. I wasn't sure about reshaping with negative numbers passed to `reshape`.

Reply:
The function is used to get the indices where the weights for a given observation is not zero. So `interval` is an array of numbers that are indices in the schedule. Where during that period of time the weights are not zero, so I can schedule the observation. That functions actually returns a list of lists, the problem with  type there is that `np.ndarray` consider both 1-dimensional or 2-dimensional arrays. So the return is a list of lists of two elements, the initial index for the interval where weights are not zero, and the final index.  Reshape by negative  is used to infer the length of the final array, I'm certain the matrix here is going to have 2 columns but not how many rows it's going to need (meaning how many non zero intervals in the weight table the observation I'm evaluating has). 

Going back to the main question, the original version of the code is here: https://stackoverflow.com/questions/24885092/finding-the-consecutive-zeros-in-a-numpy-array. Basically what it does is creating a padded version of the original vector, but changes the values when are greater than zero. than it calculate the difference between them so I get the number of values there are not zero (instead of zero as the original) and reshape the output to get the final matrix.